File: typescript/packages/core/src/vfs/adapter.ts
import type { Accessor } from '../accessor/base.ts'
import type { CommandIO } from '../commands/builtin/generic_bind/adapter.ts'
import { streamFromBytes } from '../commands/builtin/utils/wrap.ts'
import { isEnoent, isEnotdir } from '../utils/errors.ts'
import type { NativeReadOps, ReadOps, WriteOps, SearchOps, ReadBytesOp, WriteOp } from './types.ts'

export interface VFSAdapterOptions<A extends Accessor = Accessor> {
  read: ReadOps<A>
  native?: NativeReadOps<A>
  writes?: WriteOps<A>
  search?: SearchOps<A>
  local?: boolean
  isMounted?: CommandIO<A>['isMounted']
  maxGlobMatches?: number
  maxDuEntries?: number
}

/** Compose capabilities into one table for commands and filesystem ops. */
export class VFSAdapter<A extends Accessor = Accessor> {
  constructor(readonly options: VFSAdapterOptions<A>) {}

  toCommandIO(): CommandIO<A> {
    const { read, native, writes, ...settings } = this.options
    return {
      ...read,
      streamsBytes: native?.readStream === undefined,
      readStream: (a, p, i) => streamFromBytes(read.readBytes, a, p, i),
      exists: async (a, p) => {
        try {
          await read.stat(a, p)
          return true
        } catch (error) {
          if (isEnoent(error) || isEnotdir(error)) return false
          throw error
        }
      },
      local: false,
      isMounted: () => true,
      ...settings,
      ...native,
      ...writes,
    }
  }
}

/** Explicit non-atomic read/modify/write append for byte stores. */
export function appendFromRead<A extends Accessor>(
  read: ReadBytesOp<A>,
  write: WriteOp<A>,
): WriteOp<A> {
  return async (accessor, path, data) => {
    let previous: Uint8Array
    try {
      previous = await read(accessor, path)
    } catch (error) {
      if (!isEnoent(error)) throw error
      previous = new Uint8Array()
    }
    const merged = new Uint8Array(previous.length + data.length)
    merged.set(previous)
    merged.set(data, previous.length)
    await write(accessor, path, merged)
  }
}

File: python/mirage/vfs/adapter.py
from dataclasses import dataclass, field
from functools import partial

from mirage.accessor.base import Accessor
from mirage.commands.builtin.generic.du import DEFAULT_MAX_DU_ENTRIES
from mirage.commands.builtin.generic_bind.adapter import CommandIO
from mirage.commands.builtin.utils.wrap import stream_from_bytes
from mirage.types import PathSpec
from mirage.utils.glob_walk import DEFAULT_MAX_GLOB_MATCHES
from mirage.vfs.types import (IsMountedOp, NativeReadOps, ReadOps, SearchOps,
                              StatOp, WriteOps)


def _mounted(accessor: Accessor) -> bool:
    return True


async def _exists(stat: StatOp, accessor: Accessor, path: PathSpec) -> bool:
    try:
        await stat(accessor, path)
    except FileNotFoundError:
        return False
    return True


@dataclass(frozen=True, kw_only=True)
class VFSAdapter:
    """Compose resource capabilities into the command and dispatcher table.

    Args:
        read (ReadOps): required listing, byte read, and stat operations.
        native (NativeReadOps): optional accelerators; byte streaming and
            existence checks otherwise derive from the required reads.
        writes (WriteOps): independent mutations, absent by default.
        search (SearchOps | None): optional native text search.
        local (bool): whether data lives on the host filesystem.
        is_mounted (IsMountedOp): optional backend availability check.
        max_glob_matches (int | None): glob expansion ceiling.
        max_du_entries (int | None): size traversal ceiling.
    """

    read: ReadOps
    native: NativeReadOps = field(default_factory=NativeReadOps)
    writes: WriteOps = field(default_factory=WriteOps)
    search: SearchOps | None = None
    local: bool = False
    is_mounted: IsMountedOp = _mounted
    max_glob_matches: int | None = DEFAULT_MAX_GLOB_MATCHES
    max_du_entries: int | None = DEFAULT_MAX_DU_ENTRIES

    def to_command_io(self) -> CommandIO:
        """Build one table shared by commands, globbing, and filesystem ops."""
        return CommandIO(
            readdir=self.read.readdir,
            read_bytes=self.read.read_bytes,
            stat=self.read.stat,
            read_stream=self.native.read_stream
            or partial(stream_from_bytes, self.read.read_bytes),
            read_range=self.native.read_range,
            exists=self.native.exists or partial(_exists, self.read.stat),
            find=self.native.find,
            du=self.native.du,
            write=self.writes.write,
            append=self.writes.append,
            create=self.writes.create,
            mkdir=self.writes.mkdir,
            unlink=self.writes.unlink,
            rmdir=self.writes.rmdir,
            rm_r=self.writes.rm_r,
            rename=self.writes.rename,
            copy=self.writes.copy,
            dir_copy=self.writes.dir_copy,
            truncate=self.writes.truncate,
            set_attrs=self.writes.set_attrs,
            search=self.search,
            local=self.local,
            is_mounted=self.is_mounted,
            max_glob_matches=self.max_glob_matches,
            max_du_entries=self.max_du_entries,
        )
